SEP-0012.md

```diff
@@ -0,0 +1,537 @@
+                 meta: Mapping = None,
+                 unit: Union[str, u.Unit] = None,
+                 copy: bool = False):
+        """
+        This signature is inherited from NDData.
+
+        The only difference is that the wcs argument should be considered mandatory.
+        """
+
+    @abc.abstractmethod
+    def __getitem__(self, item: Union[int, slice, Iterable[Union[slice, int]]]) -> "NDCubeABC":
+        """
+        Slice the NDCube object by array index.
+
+        An ndcube object must be sliceable in a similar way to a numpy array,
+        however, support for a `step` is not required and no "fancy indexing" is allowed.
+
+        A step is prohibited due to the abiguity inherent in applying a step to the WCS
+        object; does the step mean a sampling so the coordinate axis becomes discontinuious
+        or does it mean that the size of the pixels should be increased by a factor of the step.
+
+        Fancy indexing, or
+        `advanced indexing <https://numpy.org/doc/stable/reference/arrays.indexing.html#advanced-indexing>`__
+        is not supported both due to similar issues with applying it to the WCS, as
+        well as it being poorly supported outside of numpy arrays.
+
+        When a slice is applied to a NDCube object it must be applied to the following attributes:
+
+          * data
+          * wcs
+          * uncertainty
+          * mask
+          * extra_coords
+
+        the following attributes must be copied unaltered to the returned NDCube:
+
+          * unit
+          * meta
+
+        """
+
+    @abc.abstractmethod
+    @property
+    def extra_coords(self) -> ExtraCoordsABC:
+        """
+        Coordinates not decribed by ``NDCubeABC.wcs`` which vary along one or more axes.
+        """
+
+    @abc.abstractmethod
+    @property
+    def global_coords(self) -> GlobalCoordsABC:
+        """
+        Coordinate metadata which applies to the whole cube.
+        """
+
+    @abc.abstractmethod
+    @property
+    def wcs(self) -> BaseHighLevelWCS:
+        """
+        The WCS transform for the NDCube.
+
+        If a `BaseLowLevelWCS` is provided to the constructor it should be
+        "upgraded" to a high level object.
+        """
+
+    @abc.abstractmethod
+    @property
+    def combined_wcs(self) -> BaseHighLevelWCS:
+        """
+        The WCS transform for the NDCube, including the coordinates specified in ``.extra_coords``.
+
+        This transform should implement the high level wcsapi, and have
+        `pixel_n_dim` equal to the number of array dimensions in the
+        `.NDCube`. The number of world dimensions should be equal to the
+        number of world dimensions in ``self.wcs`` and in ``self.extra_coords`` combined.
+        """
+
+    @abc.abstractmethod
+    @property
+    def array_axis_physical_types(self) -> Iterable[Tuple[str, ...]]:
+        """
+        Returns the WCS physical types associated with each array axis.
```

Comment:
Yeah I think that's a good improvement. One could argue that for a reasonably aligned grid the rate of change of one of the spatial dimensions could be vanishingly small but it would still be in this list. It still technically varies with the axis though!